aim/ext/notebook/notebook.py

```diff
@@ -0,0 +1,207 @@
+import shlex
+
+from aim.cli.configs import VERSION_NAME, UP_NAME
+import aim.ext.manager.manager as manager
+
+# Error message prefix for aim commands
+ERROR_MSG_PREFIX = b'Error:'
+
+# returned by get_execution_context
+_COLAB_EXEC_CONTEXT = "_COLAB_EXEC_CONTEXT"
+_IPYTHON_EXEC_CONTEXT = "_IPYTHON_EXEC_CONTEXT"
+_OTHER_EXEC_CONTEXT = "_OTHER_EXEC_CONTEXT"
+
+# current execution context
+_CURRENT_CONTEXT = _OTHER_EXEC_CONTEXT
+
+
+def get_execution_context():
+    """Determine the most specific context that we're in.
+    Returns:
+      _COLAB_EXEC_CONTEXT: If in Colab with an IPython notebook context.
+      _IPYTHON_EXEC_CONTEXT: If we are in an IPython notebook
+        context but not in colab (i.e. `jupyter notebook`)
+        line).
+      _OTHER_EXEC_CONTEXT: Otherwise (e.g., by running a Python script at the
+        command-line or using the `ipython` interactive shell).
+    """
+    # In Colab, the `google.colab` module is available, but the shell
+    # returned by `IPython.get_ipython` does not have a `get_trait`
+    # method.
+
+    # imports are dynamic, since some modules are not available for all contexts
+    try:
```

Comment:
code duplication
better to rewrite with this:
```    
ipython = None
try:
        import IPython
except ImportError:
        pass
else:
        ipython = IPython.get_ipython()
```

and the have :
```
if ipython is not None and 'google.colab' in str(ipython):
    ...
elif ipython is not None and ipython.has_trait("kernel"):
    ...
else:
    ...
```

Reply:
Here is a stupid moment, it need to import colab to check ths is in colab, if it can not import colab, should not rase exception because maybe in IPython, so the nested try catch block is requried.